docs(framework): CatBoost guide (#2716)
bentoml/BentoML: docs/source/frameworks/catboost.rst

       1.312e-01, 2.776e-01, 1.890e-01, 7.283e-02, 3.184e-01, 8.183e-02]]))

Building a Service using CatBoost
--------------------------------

.. seealso::

   :ref:`Building a Service <concepts/service:Service and APIs>`: more information on creating a prediction service with BentoML.

.. code-block:: python

   import bentoml

   import numpy as np

   from bentoml.io import NumpyNdarray

   runner = bentoml.catboost.get("catboost_cancer_clf:latest").to_runner()

   svc = bentoml.Service("cancer_clf", runners=[runner])

   @svc.api(input=NumpyNdarray(), output=NumpyNdarray())
   def classify_cancer(input: np.ndarray) -> np.ndarray:
      # returns sentiment score of a given text
      res = runner.predict.run(input)
      return res

When constructing a :ref:`bentofile.yaml <concepts/bento:Bento Build Options>`,
there are two ways to include CatBoost as a dependency, via ``python`` or
``conda``:

.. tab-set::

   .. tab-item:: python

      .. code-block:: yaml

         python:
         - catboost

   .. tab-item:: conda

      .. code-block:: yaml

         conda:
           channels:
           - conda-forge
           dependencies:
           - catboost

> Using Runners

Reviewer: add a note here that we only support `predict` and `predict_proba` as a signature for now.

`staged_predict` is not yet supported.

-------------

.. seealso::

   See :ref:`concepts/runner:Using Runners` doc for a general introduction to the Runner concept and its usage.

A runner for a Booster is created like so:

.. code-block:: python

   bentoml.catboost.get("model_name:model_version").to_runner()

``runner.predict.run`` is generally a drop-in replacement for ``model.predict``. However, while it
is possible to pass a ``Pool`` as input, BentoML does not support adaptive batching in that case.
It is therefore recommended to use a NumPy ``ndarray`` or Pandas ``DataFrame`` as input instead.

.. note
   BentoML only support ``predict`` and ``predict_proba`` as a signature for now. ``staged_predict`` is not yet supported.

Using GPU
---------

If there is a GPU available, the CatBoost Runner will automatically use ``task_type=GPU`` by default.
This can be disabled by using the
:ref:`BentoML configuration file <guides/configuration:Configuring BentoML>` to disable Runner GPU
access:

.. code-block:: yaml
   runners:
      # resources can be configured at the top level
      resources:
         nvidia.com/gpu: 0
      # or per runner
      my_runner_name:
         resources:
             nvidia.com/gpu: 0

> Adaptive batching 

Reviewer 2: Link to batching article

cc https://github.com/bentoml/BentoML/pull/2708/files#r918338794

========
CatBoost
========

CatBoost is a machine learning algorithm that uses gradient boosting on decision trees. It is available as an open source library.
To learn more about CatBoost, visit their `documentation <https://catboost.ai/en/docs/>`_.

BentoML provides native support for `CatBoost <https://github.com/catboost/catboost>`_, and this guide provides an overview of how to use BentoML with CatBoost.

Saving a trained CatBoost model
--------------------------------

> In this example, we will be training a new model using UCI's `breast cancer dataset <https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)>`_.

Reviewer: ```suggestion
In this example, we will train a new model using UCI's `breast cancer dataset <https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)>`_.
```
```suggestion
In this example, we will be training a new model using UCI's `breast cancer dataset <https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(diagnostic)>`_.
```

.. code-block:: python

   import bentoml

   import catboost as cbt

   from sklearn.datasets import load_breast_cancer

   cancer = load_breast_cancer()

   X = cancer.data
   y = cancer.target

   model = cbt.CatBoostClassifier(
       iterations=2,
       depth=2,
       learning_rate=1,
       loss_function="Logloss",
       verbose=False,
   )

   # train the model
   model.fit(X, y)

> After training, use :obj:`~bentoml.catboost.save_model` to save the model instance to BentoML model store.

Reviewer: ```suggestion
Use :obj:`~bentoml.catboost.save_model()` to save the model instance to BentoML model store:
```
```suggestion
After training, use :obj:`~bentoml.catboost.save_model` to save the model instance to BentoML model store.
```

.. seealso::

   :ref:`Building a Service <concepts/service:Service and APIs>`: more information on creating a prediction service with BentoML.

.. code-block:: python

   import bentoml

   import numpy as np

   from bentoml.io import NumpyNdarray

   runner = bentoml.catboost.get("catboost_cancer_clf:latest").to_runner()

   svc = bentoml.Service("cancer_clf", runners=[runner])

   @svc.api(input=NumpyNdarray(), output=NumpyNdarray())
   def classify_cancer(input: np.ndarray) -> np.ndarray:
      # returns sentiment score of a given text
      res = runner.predict.run(input)
      return res

When constructing a :ref:`bentofile.yaml <concepts/bento:Bento Build Options>`,
there are two ways to include CatBoost as a dependency, via ``python`` or
``conda``:

.. tab-set::

   .. tab-item:: python

      .. code-block:: yaml

         python:
         - catboost

   .. tab-item:: conda

      .. code-block:: yaml

         conda:
           channels:
           - conda-forge
           dependencies:
           - catboost

Using Runners
-------------

.. seealso::

   See :ref:`concepts/runner:Using Runners` doc for a general introduction to the Runner concept and its usage.

> A runner for a Booster is created like so:
> 
> .. code-block:: python
> 
>    bentoml.catboost.get("model_name:model_version").to_runner()

Reviewer: ```suggestion
A CatBoost :obj:`~bentoml.Runner` can be created as follow:

.. code-block:: python

   runner = bentoml.catboost.get("cancer_detection:latest").to_runner()
```

> ``runner.predict.run`` is generally a drop-in replacement for ``model.predict``. However, while it
> is possible to pass a ``Pool`` as input, BentoML does not support adaptive batching in that case.
> It is therefore recommended to use a NumPy ``ndarray`` or Pandas ``DataFrame`` as input instead.

Reviewer: ```suggestion
``runner.predict.run`` is generally a drop-in replacement for ``model.predict``.

While a `Pool <https://catboost.ai/en/docs/concepts/python-reference_pool>`_ input type can be passed to a CatBoost Runner, BentoML will not support adaptive batching due to current implementation.

To use adaptive batching feature from BentoML, we recommend our users to use either NumPy ``ndarray`` or Pandas ``DataFrame`` instead.
```

.. note

>    BentoML only support ``predict`` and ``predict_proba`` as a signature for now. ``staged_predict`` is not yet supported.

Reviewer: ```suggestion
   Currently ``staged_predict`` callback is not yet supported.
```
```suggestion
   BentoML only support ``predict`` and ``predict_proba`` as a signature for now. ``staged_predict`` is not yet supported.
```

Using GPU
---------

> If there is a GPU available, the CatBoost Runner will automatically use ``task_type=GPU`` by default.
> This can be disabled by using the
> :ref:`BentoML configuration file <guides/configuration:Configuring BentoML>` to disable Runner GPU
> access:

Reviewer: ```suggestion
CatBoost Runner will automatically use ``task_type=GPU`` if GPU is available on the system.
This behaviour can then be disabled via :ref:`BentoML's configuration<guides/configuration:Configuring BentoML>`:
access:
```
```suggestion
If there is a GPU available, the CatBoost Runner will automatically use ``task_type=GPU`` by default.
This can be disabled by using the
:ref:`BentoML configuration file <guides/configuration:Configuring BentoML>` to disable Runner GPU
access:
```

Reviewer 2: I think the old wording was fine?

.. code-block:: yaml

   runners:
      # resources can be configured at the top level
      resources:
         nvidia.com/gpu: 0
      # or per runner
      my_runner_name:
         resources:
             nvidia.com/gpu: 0

Adaptive batching 
~~~~~~~~~~~~~~~~~

.. seealso::

   :ref:`guides/batching:Adaptive Batching`: a general introduction to adaptive batching in BentoML.

CatBoost's ``model.predict`` supports taking batch input for inference. This is disabled by

> default, but can be enabled using the appropriate signature when saving your booster.

Reviewer: ```suggestion
default, but can be enabled using the appropriate signature when saving a Booster object.
```

.. note

   BentoML does not currently support adaptive batching for ``Pool`` input. In order to enable

>    batching, use either a NumPy ``ndarray`` or a Pandas ``DataFrame`` instead.

Reviewer 2: The reason I have this in both sections is because a user may skim either; we could probably put something like "as stated above" in though.

``conda``:

.. tab-set::

   .. tab-item:: python

      .. code-block:: yaml

         python:
         - catboost

   .. tab-item:: conda

      .. code-block:: yaml

         conda:
           channels:
           - conda-forge
           dependencies:
           - catboost

Using Runners
-------------

.. seealso::

   See :ref:`concepts/runner:Using Runners` doc for a general introduction to the Runner concept and its usage.

A CatBoost :obj:`~bentoml.Runner` can be created as follow:

.. code-block:: python

   runner = bentoml.catboost.get("catboost_cancer_clf:latest").to_runner()

``runner.predict.run`` is generally a drop-in replacement for ``model.predict``.

While a `Pool <https://catboost.ai/en/docs/concepts/python-reference_pool>`_ input type can be passed to a CatBoost Runner, BentoML will not support adaptive batching due to current implementation.

To use adaptive batching feature from BentoML, we recommend our users to use either NumPy ``ndarray`` or Pandas ``DataFrame`` instead.

.. note::

   Currently ``staged_predict`` callback is not yet supported with :code:`bentoml.catboost`.

Using GPU
---------

> CatBoost Runner will automatically use ``task_type=GPU`` if GPU is available on the system.

Reviewer 2: ```suggestion
CatBoost Runners will automatically use ``task_type=GPU`` if a GPU is detected.
```

> This behaviour can then be disabled via :ref:`BentoML's configuration<guides/configuration:Configuring BentoML>`:

Reviewer 2: ```suggestion
This behavior can be disabled using the :ref:`BentoML configuration file<guides/configuration:Configuring BentoML>`:
```

access:

.. code-block:: yaml

   runners:
      # resources can be configured at the top level
      resources:
         nvidia.com/gpu: 0
      # or per runner
      my_runner_name:
         resources:
             nvidia.com/gpu: 0

Adaptive batching 
~~~~~~~~~~~~~~~~~

.. seealso::

   :ref:`guides/batching:Adaptive Batching`: a general introduction to adaptive batching in BentoML.

CatBoost's ``model.predict`` supports taking batch input for inference. This is disabled by
default, but can be enabled using the appropriate signature when saving your model.

.. note::

   BentoML does not currently support adaptive batching for ``Pool`` input. In order to enable
   batching, use either a NumPy ``ndarray`` or a Pandas ``DataFrame`` instead.

.. code-block:: python

>    bento_model = bentoml.catboost.save_model("catboost_cancer_clf", model, signatures={"predict": {"batchable": True}})

Reviewer: ```suggestion
   bento_model = bentoml.catboost.save_model(
    "catboost_cancer_clf", model, signatures={"predict": {"batchable": True}}
)
```
